sys/fido2/ctap/ctap.c

```diff
@@ -0,0 +1,1855 @@
+    if (_state.rk_amount_stored == 0) {
+        for (uint16_t i = 0; i < allow_list_len; i++) {
+            ret = _ctap_decrypt_rk(&rks[index], &allow_list[i].cred_id);
+            if (ret == CTAP2_OK) {
+                if (memcmp(rks[index].rp_id_hash, rp_id_hash,
+                           SHA256_DIGEST_LENGTH) == 0) {
+                    index++;
+                }
+            }
+        }
+    }
+
+    for (int i = 0; i < _state.rk_amount_stored; i++) {
+        uint16_t page_num = _get_flashpage_number_of_rk(i);
+        uint16_t offset_into_page = _get_offset_of_rk_into_flashpage(i);
+
+        if (offset_into_page == 0) {
+            fido2_ctap_mem_read(CTAP_FLASH_RK_START_PAGE + page_num, page);
+        }
+
+        memcpy(&rk, page + offset_into_page, sizeof(rk));
+
+        /* search for rk's matching rp_id_hash */
+        if (memcmp(rk.rp_id_hash, rp_id_hash, SHA256_DIGEST_LENGTH) == 0) {
+            if (allow_list_len == 0) {
+                memcpy(&rks[index], &rk, sizeof(rk));
+                index++;
+            }
+            else {
+                /* if allow list is present, also check that cred_id is in list */
+                for (size_t j = 0; j < allow_list_len; j++) {
+                    if (memcmp(allow_list[j].cred_id.id, rk.cred_desc.cred_id,
+                               sizeof(rk.cred_desc.cred_id)) == 0) {
+                        memcpy(&rks[index], &rk, sizeof(rk));
+                        index++;
+                        break;
+                    }
+                    else {
+                        /* no match with stored key, try to decrypt */
+                        ret = _ctap_decrypt_rk(&rks[index],
+                                               &allow_list[j].cred_id);
+                        if (ret == CTAP2_OK) {
+                            if (memcmp(rks[index].rp_id_hash, rk.rp_id_hash,
+                                       SHA256_DIGEST_LENGTH) == 0) {
+                                index++;
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        if (index >= rks_len) {
+            break;
+        }
+    }
+
+    /**
+     * Sort in descending order based on creation time. Credential with the
+     * most recent (highest) creation time will be first in list.
+     */
+    if (index > 0) {
+        qsort(rks, index, sizeof(ctap_resident_key_t), fido2_ctap_utils_cred_cmp);
+    }
+
+    return index;
+}
+
+/**
+ * overwrite existing key if equal, else find free space.
+ *
+ * The current official CTAP spec does not have credential management yet
+ * so rk's can't be deleted, only overwritten => we can be sure that there are
+ * no holes when reading keys from flash memory
+ */
+static int _save_rk(ctap_resident_key_t *rk)
+{
+    bool equal = false;
+    int ret;
+    uint16_t page_num = 0, offset_into_page = 0;
+    uint8_t page[FLASHPAGE_SIZE] = { 0 };
```

Comment:
`FLASHPAGE_SIZE` can be pretty big (8k on SAM D5x/E5x), it's risky putting that on the stack.

You could have saved yourself a lot of hassle by using the [MTD API](https://doc.riot-os.org/group__drivers__mtd.html) that abstracts away the low-level flash handling and comes with it's own flashpage sized work buffer.

Reply:
Oh wow ... I have overlooked this API completely. I will port the code to use this API 👍